### ISSUE TITLE
Difficult to use React components with Highcharts formatters.

### ISSUE THREAD

**Issue by jon-a-nygaard** [highcharts/highcharts-react#23]
I have received feedback from a user asking how they can render React components in a Highcharts formatter function.
This proved to be quite difficult and we should try to somehow make it more user friendly.

I made one example where I am rendering a component with a react router in xAxis.labels.formatter using a helper function named `formatter`: [Gist](https://gist.github.com/jon-a-nygaard/7d0253b2c73ae634d5804d6794a67c0c)
The solution was complicated by a bug in Highcharts.fireEvent, and that the Highcharts formatters expects a string in return.

One way I believe this can perhaps be simplified is that the Highcharts formatters can be modified to handle dom elements as return values.
Still this will not solve the full issue, as the Highcharts formatters will still not be able to handle a React component. So in my opinion there will likely be a need for some sort of helper as middleware.

**Comment by Anwardo** [comment]
Any update on this? Being able to render dom nodes inside formatters would make a huge difference. Right now it's really frustrating to properly add event handlers to elements inside a formatter function.

**Comment by ppotaczek** [comment]
Hello,

Unfortunately, there is currently no progress in this topic. As the autor of this issue rightly noticed, the implementation would be complicated and would require some changes in the Highcharts core.

However, I think that I can suggest a slightly different approach. React v16 provided portals:

> Portals provide a first-class way to render children into a DOM node that exists outside the DOM hierarchy of the parent component.

which allow us to add React components for example to `xAxis` labels. Please check the following example:

**Live demo:** https://codesandbox.io/s/1o5y7r31k3
**Docs:** https://reactjs.org/docs/portals.html

Best regards!

**Comment by Anwardo** [comment]
Hi, thanks for your detailed response, provided examples and insights! I considered using Portals but didn't figure out how to actually bring them into the chart yet. Very interesting solution.

I managed to utilize renderToString while binding a listener to a generated id for my use case for now, but might look into using your provided solution at a later stage.

**Comment by jon-a-nygaard** [comment]
Related issue in highcharts repository https://github.com/highcharts/highcharts/issues/7755

**Comment by ksdelacruz** [comment]
I want to +1 on this feature. Currently, I'm trying to add React Router Links on my heatmap labels in order to load more detailed charts though since strings are the only accepted returns for formatters, I can't get to make router links work. Mimicking links using the <a> tags on labels would make the page reload which is different from the intended SPA behavior.

**Comment by ppotaczek** [comment]
Hi @ksdelacruz, Did you try to use the solution with portals?

**Comment by ksdelacruz** [comment]
@ppotaczek I can't get to work the solution (or at least I can see how to incorporate it) on my development, because unlike the given example that access the ticks/label of x-Axis, I want to access the actual labels (points) of the series. Anyway, I just utilized the "onClick" event of the chart to save the point/label and just re-render a button with `<Link>` tag outside the  heatmap.

**Comment by ppotaczek** [comment]
@ksdelacruz, 

By portals you can add React components to all HTML chart's elements. Please check the example with data labels: https://codesandbox.io/s/highcharts-react-demo-v9dns

Best regards!

**Comment by ksdelacruz** [comment]
@ppotaczek Thank you very much for your help!

**Comment by stovmascript** [comment]
Don't know if this is any help, but we're currently using [`renderToStaticMarkup`](https://reactjs.org/docs/react-dom-server.html#rendertostaticmarkup) to enable using JSX in the formatting function:

```js
import { renderToStaticMarkup } from 'react-dom/server'

// ...
  return (
    <HighchartsReact
      options={{
        tooltip: {
          formatter() {
            return renderToStaticMarkup(
              <span style={{ color: 'red' }}>
                {/* ... */}
```

**Comment by ksdelacruz** [comment]
@ppotaczek I would like to ask for additional help. I tried converting the implementation of your given solution into React Hooks but there's a problem I encountered and I can't pinpoint the problem right now. The portals do work but only on the first row only of the heatmap. Hovering on the other rows would release an error 
> Uncaught TypeError: b.onMouseOver is not a function
    at HTMLDivElement.e

**Comment by ppotaczek** [comment]
Hi @ksdelacruz,

Could you reproduce that issue in some online code editor? You can edit the live example that I provided previously.

**Comment by ksdelacruz** [comment]
@ppotaczek Here's the example of the code that I've been talking to. https://codesandbox.io/s/highcharts-react-demo-w4fix

Kindly refer to the console for further clarifications. Thank you very much!

**Comment by ppotaczek** [comment]
Hi @ksdelacruz, Thank you for the live demo!

The problem is that the chart re-renders after adding components from the portal, please check the order of functions in this example: https://codesandbox.io/s/highcharts-react-demo-tqzrr

You can prevent re-rendering, for example by setting `allowChartUpdate` option to false:

    <HighchartsReact
        ...
        allowChartUpdate={false}
    />

**Live example:** https://codesandbox.io/s/highcharts-react-demo-nvve6

Best regards!

**Comment by ksdelacruz** [comment]
@ppotaczek Again, thank you very much for assisting me!

**Comment by nmdhulipud** [comment]
Are we still good here to use Highchart with react, I am starting a new project here. Let me know your experience with Highcharts so far. 
In my mind, I want to use recharts.js which is pretty much compatible with React js, my friend insists on using highcharts. 

Thanks

**Comment by cstoddart** [comment]
I tried both portals and jsx to string solutions. I decided to go with a jsx to string solution similar to @stovmascript 's suggestion. This works with either `renderToString()` or `renderToStaticMarkup()`.

```js
import { renderToString } from 'react-dom/server';
import { Tooltip } from './CustomTooltip';

// ...
  return (
    <HighchartsReact
      options={{
        tooltip: {
          formatter: function() { // you can't use an arrow function here if you want to access the tooltip data off of 'this'
            const { x, y, point, series } = this;
            return renderToString(<Tooltip {...{ x, y, point, series }} />);
          },
          {/* ... */}

**Comment by dankremniov** [comment]
Both `renderToString` and `renderToStaticMarkup` work quite well. Obviously, they strip out all event handlers from components, and, thus, for tooltips with actions - usage of event bubbling is the only option.

However, since Highcharts v9.0.0, rendering of React components in the tooltip can be quite straightforward using React portals (thank you for the idea, @ppotaczek). In the previous version, tooltip DOM content was completely rerendered on each refresh regardless of whether it has changed or not (which was making the implementation using portals quite complicated).

I published POC for v9.0.0 using React portals here: [https://gist.github.com/dankremniov/a9a6b969e63dfc4f0f83e6f82b82eb4f](https://gist.github.com/dankremniov/a9a6b969e63dfc4f0f83e6f82b82eb4f). @ppotaczek, I would love to hear your feedback and any suggestion.

**Comment by ppotaczek** [comment]
Hi @dankremniov,

Thanks for your comments.

Your code looks really well! The only thing I can see to improve is chart updating. You don't need to merge options, it is enough to apply only the new ones:

      chart.update(
        tooltip: {
          formatter,
          useHTML: true
        }
      );
      
Best regards!

**Comment by dankremniov** [comment]
@ppotaczek, thank you 👍 I can add the link to the codesandbox to the corresponding FAQ section of the readme if you think it is worth it.

**Comment by ppotaczek** [comment]
Yes, it is worth it 👍  good idea!

**Comment by ppotaczek** [comment]
Hi @plotka,

I just checked the example and it also works for version 3.0.0, maybe you didn't update React version?

**Live example:** https://codesandbox.io/s/highcharts-react-demo-forked-qk6cb?file=/demo.jsx

**Comment by plotka** [comment]
Hi @ppotaczek. You are right, the React version needed to be updated. However, I tried to convert it into a functional component and for some reason, the labels are not being rendered. Do you think you could help? Thanks.

**Live example**: https://codesandbox.io/s/highcharts-react-demo-forked-sh3qt?file=/demo.jsx

**Comment by ppotaczek** [comment]
Hello @plotka, I converted the example into a functional component with small improvements here:
https://codesandbox.io/s/highcharts-react-demo-forked-e1v82?file=/demo.jsx

**Comment by maciej-gurban** [comment]
Should I be seeing those custom labels in the above demo @ppotaczek? This is what renders for me in the linked code sandbox. Can't see the "React" word being rendered anywhere
![Screen Shot 2021-03-22 at 16 09 30](https://user-images.githubusercontent.com/3663640/112002935-14928a00-8b29-11eb-81ec-3e6ede1fb1f5.png)

**Comment by ppotaczek** [comment]
Hello @maciej-gurban, I recommend you to check the example from @dankremniov 
Please let me know in case of any questions.

**Comment by maciej-gurban** [comment]
I've reached out to him and got help in adjusting that solution to our use case, thanks. Do you think you could incorporate part of his solution into highcharts-react documentation, or even making it part of the library?

**Comment by ppotaczek** [comment]
Ok, great! Every case of using portals is slightly different, so adding something like this to the library is not a good idea, but I will add an example with tooltip to the docs.

**Comment by alvarotrigo** [comment]
What would be the main benefit of using `createPortal` instead of `ReactDOMServer.renderToStaticMarkup`?
Is `ReactDOMServer` increasing the bundle size? 
Does `ReactDOMServer.renderToStaticMarkup` has any performance downsides?

@dankremniov any idea?

**Comment by dankremniov** [comment]
@alvarotrigo, I mentioned one the reasons [here](https://github.com/highcharts/highcharts-react/issues/23#issuecomment-774694996) previously. Another issue is that with `ReactDOMServer.renderToStaticMarkup` you will not be able to consume context within the tooltip unless you explicitly wrap it into a provider. Otherwise, if you need relatively simple tooltip without event handling or context - `renderToStaticMarkup` is a good option.

**Comment by alvarotrigo** [comment]
@dankremniov yeah I read those, i was more concerned about performance and bundle size downsides. Does ReactDOMServer increase the bundle size at all?

**Comment by dankremniov** [comment]
@alvarotrigo, I was not concerned about bundle sizes at that time, so, unfortunately, I cannot comment on that. I have not noticed any significant different in rendering performance.

**Comment by ashleyryan** [comment]
Does anyone have an example of a tooltip using a portal? I see the label example, but there's no tooltip dom element available at render time as far as I'm aware. I would prefer not to use `renderToStaticMarkup` since it's not part of `react-dom`

**Comment by alvarotrigo** [comment]
> I would prefer not to use `renderToStaticMarkup` since it's not part of `react-dom`

It increases bundle size?

**Comment by ashleyryan** [comment]
> > I would prefer not to use `renderToStaticMarkup` since it's not part of `react-dom`
> 
> It increases bundle size?

It's a separate package so ... yeah, I'm sure it will, though I haven't investigated by how much. I'm also building this as a reusable component and so don't want to lose events, which you'll lose when rendering to plain old html

**Comment by ashleyryan** [comment]
I got something half working, built on top of the example above and using one of the demos from the documentation

It's a really bad frankenstein of portal code plus rendering using the renderer - I need to be able to resize the tooltip I'm trying to render content into.

https://codesandbox.io/s/highcharts-react-demo-forked-f83ci?file=/demo.jsx

**Comment by arnav-dt** [comment]
renderToString works for simple dom nodes like divs and spans. However button + onclick handler fails inside tooltip useHTML formatter. Any idea how to get around this ? All onclicks fail
![image](https://user-images.githubusercontent.com/80442805/148761038-39888a53-d3e4-4215-9dab-e2f98438b528.png)

Sidenote: Researched a lot on this and found https://codesandbox.io/s/highcharts-react-demo-gwk4k however this does not work with typescript as `this` is implicity any and I would need change tsconfig for the entire app.
Like http://jsfiddle.net/BlackLabel/L6hygs35/ used to work but now fails @ppotaczek

**Comment by Anwardo** [comment]
> renderToString works for simple dom nodes like divs and spans. However button + onclick handler fails inside tooltip useHTML formatter. Any idea how to get around this ? All onclicks fail ![image](https://user-images.githubusercontent.com/80442805/148761038-39888a53-d3e4-4215-9dab-e2f98438b528.png)
> 
> Sidenote: Researched a lot on this and found https://codesandbox.io/s/highcharts-react-demo-gwk4k however this does not work with typescript as `this` is implicity any and I would need change tsconfig for the entire app. Like http://jsfiddle.net/BlackLabel/L6hygs35/ used to work but now fails @ppotaczek

Indeed, assigning an event handler does not work with renderToString. The workaround I'm using is assigning an ID, class or data attribute to your button and adding an event handler to the selector in useEffect. Not 'react-y' but it works.

**Comment by ppotaczek** [comment]
Hi @arnav-dt, 
To add the `onclick` attribute, you need to add it to allowed attributes first:
`Highcharts.AST.allowedAttributes.push('onclick')` 
Example: http://jsfiddle.net/BlackLabel/9w1za65g/

It will work with React and TypeScript in the same way, only type `this` in the formatter function to `Highcharts.TooltipFormatterContextObject` 
API: https://api.highcharts.com/class-reference/Highcharts#.TooltipFormatterCallbackFunction

You can also think about portals: https://github.com/highcharts/highcharts-react#how-to-add-react-component-to-a-charts-element

**Comment by arnav-dt** [comment]
Thank you @ppotaczek and @Anwardo for your inputs I am able to fire an event using buttons onclick and listen to it in my react code. This will work well as a workaround for my usecase (button inside tooltip).

I tried using portals/ReactDOM.render() but the issue with them is that I need an HTML element to latch onto as the second parameter. For tooltips the tooltip HTML DOM nodes are created only after mouse hovers any of the points/bars. Until then there is no DOM node for react render code inside of.

**Comment by ppotaczek** [comment]
I am glad to hear that. According to portals, please check a comment and code example in a comment from this thread: https://github.com/highcharts/highcharts-react/issues/23#issuecomment-774694996

**Comment by MaheshMohan3856** [comment]
> Hi @arnav-dt, To add the `onclick` attribute, you need to add it to allowed attributes first: `Highcharts.AST.allowedAttributes.push('onclick')` Example: http://jsfiddle.net/BlackLabel/9w1za65g/
> 
> It will work with React and TypeScript in the same way, only type `this` in the formatter function to `Highcharts.TooltipFormatterContextObject` API: https://api.highcharts.com/class-reference/Highcharts#.TooltipFormatterCallbackFunction
> 
> You can also think about portals: https://github.com/highcharts/highcharts-react#how-to-add-react-component-to-a-charts-element

Can attach a example code for this in Reactjs

**Comment by arnav-dt** [comment]
Writing some rough code which worked for me. I don't have the exact code with me right now.
Add onclick to Highcharts
```
Highcharts.AST.allowedAttributes.push('onclick');
```

In Highcharts options use the tooltip property like so:
```
tooltip: {
    borderColor: 'transparent',
    borderRadius: 0,
    borderWidth: 0,
    className: 'charts-tooltip',
    outside: true,
    shadow: false,
    shape: 'square',
    shared: true,
    useHTML: true,
    formatter(): string {
      const scope = this as any;
      const tooltipInner = scope.points.reduce(
        (
          s: string,
          point: {
            x: string;
            y: number;
          },
        ) => {
          const pointerStyle = 'style="cursor: pointer"';
          const onClickAction = `onclick="document.dispatchEvent(new CustomEvent('my-event-${point.series.options.chartId}', { 'detail': { 'xVal': '${point.x}'}' } }));"`;
          const divAttributes = `${pointerStyle} ${onClickAction}`;
          return `${s}
                  <div class="charts-tooltip-inner"  ${divAttributes}>
                    <span>
                      Your other tooltip content goes here
                    </span>
                  </div>`;
        },
        `<div class="charts-tooltip-header">${scope.x}</div>`,
      );
      return `<div class="charts-tooltip-wrap">${tooltipInner}</div>`;
    },
  },
```

And an event listener in your react code to listen to the onclick
```
  useEffect(() => {
    const handleClick = ({
      detail: { xVal },
    }: any) => {
      console.log(xVal);
    };

    document.addEventListener(
      `my-event-${chartId}`,
      handleClick,
    );

    return () => {
      document.removeEventListener(
        `my-event-${chartId}`,
        handleClick,
      );
    };
  }, []);
```
However I highly recommend you go by the portal way since the code becomes much more maintainable https://github.com/highcharts/highcharts-react/issues/23#issuecomment-774694996

**Comment by aAntak** [comment]
Did anyone had a chance to make tooltip size dynamic  & adjust to the content size when using portals? 
![image](https://user-images.githubusercontent.com/22105472/192013136-97828602-e053-45f3-a7be-f089b3d1f507.png)

https://codesandbox.io/s/highcharts-react-tooltip-forked-5rc5xz

**Comment by arnav-dt** [comment]
Hello @aAntak You can render any custom react component using portals. The size can dynamically adjust based on the css of the custom react tooltip container used

**Comment by boris773** [comment]
@aAntak I do not think it's possible. As far as I'm aware CSS will not help here. If we would have FIXED content on every point of the chart, we could specify width, but currently, if the content might be larger/smaller on every point, the tooltip will not resize.

**Comment by dankremniov** [comment]
@aAntak, there is a [possible solution](https://github.com/highcharts/highcharts-react/issues/23#issuecomment-774694996) above in the thread.

**Comment by aAntak** [comment]
@arnav-dt @boris773 @dankremniov  thanks for answering!

I used that forked solution in my first question. If you hover on first point, and then on the second one you will see that the content overflows the tooltip container.  https://codesandbox.io/s/highcharts-react-tooltip-forked-5rc5xz

**Comment by ppotaczek** [comment]
@aAntak, You need to adjust dimensions of the box SVG element. Please check this thread: https://www.highcharts.com/forum/viewtopic.php?f=9&t=49412

**Comment by aAntak** [comment]
@ppotaczek  thanks for reply,

Unfortunately, this does have "blinking & "jumping" effect. As far as I understand this happens because we are correcting tooltip size on the next render.  https://codesandbox.io/s/highcharts-custom-tooltip-forked-9ji67s?file=/features/HighchartsTooltip.ts  .

**Comment by ppotaczek** [comment]
@aAntak, Yes, it is really problematic to use portals with tooltip due to it's under the hood SVG logic. Another solution may be to build your own custom HTML tooltip, bound with the `mouseOver` point's event. Please check the below example.

**Live example:** https://codesandbox.io/s/highcharts-react-demo-fork-2tdxbb?file=/demo.jsx
**API Reference:** https://api.highcharts.com/highcharts/plotOptions.line.point.events.mouseOver

**Comment by Angusross111** [comment]
I found the last solution using the custom HTML tooltip bound with the mouseOver event has serious performance issues with large amounts of data due to the whole chart rerendering each time the mouseOver event fires due to the setState.
https://codesandbox.io/s/highcharts-react-demo-fork-forked-bwtmm3?file=/demo.jsx

I found a solution by using reacts useImperativeHandle to move the Tooltip setState to its own component, so it is the only thing that needs rerendering, not the whole chart.
https://codesandbox.io/s/highcharts-react-demo-fork-forked-29pvvi?file=/demo.jsx

**Comment by Revanth12-creator** [comment]
I am facing highcharts_react_official_1 is not defined in React . Any solution ?

**Comment by ppotaczek** [comment]
Hi @Revanth12-creator, Please create a new issue with a more detailed description of the problem.

**Comment by abhirocks550** [comment]
I am using JEST and RTL for writing unit tests. I can see charts being rendered inside tests when I check below assertion it is true 
`expect(container.getElementsByClassName("highcharts-point").length).toBeGreaterThan(0)` 

when I trigger event on the **highcharts-point** I also see **highcharts-series-hover** also gets applied which means hover was successful but the issue is **formatter** function is not triggered from tooltip options for highcharts due to which tooltip is not displayed and I am not able to test further assertions. can someone help what am I missing here?

**Comment by vivekd005** [comment]
> Hello @plotka, I converted the example into a functional component with small improvements here: https://codesandbox.io/s/highcharts-react-demo-forked-e1v82?file=/demo.jsx

Hi @ppotaczek, was there any specific reason for not using callback for chart context as in other class-based examples? 
I used the callback method and faced the same issue as @plotka. I was able to console everything but for some reason labels were not rendering. Also, I was facing errors with the improvements you made for the Functional Component. It's giving me an error, **Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.** I

**Comment by KamilKubik** [comment]
From v4 onward ([@highcharts/react](https://www.npmjs.com/package/@highcharts/react)), you can render different chart elements using dedicated React components (see the [full list here](https://www.highcharts.com/docs/react/components/chart-elements/title)).

Example code (for more detailed information, you can also refer to the [component-wrapping](https://www.highcharts.com/docs/react/component-wrapping) documentation):
```
function CustomTitle() {
  return <div>Custom title</div>;
}

export default function App() {
  return (
    <Chart>
      <Title>
        <CustomTitle />
      </Title>
      <LineSeries data={[3, 1, 4, 1, 5]} />
    </Chart>
  );
}
```

For Highcharts React v3 and below, the recommended approach is to use [portals](https://react.dev/reference/react-dom/createPortal), as described [here](https://www.npmjs.com/package/highcharts-react-official#how-to-add-react-component-to-a-charts-element). However, since this isn't built into the integration, there are other potential approaches discussed in this thread - which one to go with may depend on your specific needs.

Closing as inactive for a while and resolved in the newest version. However, feel free to reopen, comment, or provide more context. Also, I'm happy to help if you struggle with any of the v3 implementations.
